LinearRegression/LinearRegressionBackend/MLNeuralNetwork/NeuralNetwork.cs

```diff
@@ -21,48 +21,55 @@ public Vector<double> PropagateForward(Vector<double> inputData)
 
             foreach (Layer layer in _Layers)
             {
-                activations = layer.Activations(activations);
+                layer.PropagateForward(activations);
+                activations = layer.Activations;
             }
 
             return activations;
         }
 
-        public void PropagateBackward(Matrix<double> actualValue, Matrix<double> expectedValue, double learningRate)
-        {
-            Debug.Assert(actualValue.RowCount == expectedValue.RowCount);
-            Debug.Assert(actualValue.ColumnCount == expectedValue.ColumnCount);
-            throw new NotImplementedException();
-        }
-
-        private Vector<double> CostDerivatives(Matrix<double> actualValue, Matrix<double> expectedValue)
+        public void PropagateBackward(
+            List<Matrix<double>> weightedSums,
+            List<Matrix<double>> activations,
+            Matrix<double> actualValues,
+            Matrix<double> expectedValues,
+            double learningRate
+        )
         {
             Layer lastLayer = _Layers[_Layers.Count - 1];
-            Vector<double> derivatives = Vector<double>.Build.Dense(lastLayer.NeuronCount);
 
-            for (int i = 0; i < actualValue.RowCount; i++)
-            {
-                // pass
-            }
+            Debug.Assert(actualValues.RowCount == lastLayer.NeuronCount);
+            Debug.Assert(actualValues.RowCount == expectedValues.RowCount);
+            Debug.Assert(actualValues.ColumnCount == expectedValues.ColumnCount);
+
+            Vector<double> cumulativeDerivatives = CostDerivatives(actualValues, expectedValues);
 
             throw new NotImplementedException();
         }
 
-        private Vector<double> ActivationDerivatives(Vector<double> actualValue, Vector<double> expectedValue)
+        private Vector<double> CostDerivatives(Matrix<double> actualValues, Matrix<double> expectedValues)
+        {
+            Vector<double> derivatives = Vector<double>.Build.Dense(actualValues.RowCount, 1);
+            derivatives = (actualValues - expectedValues) * derivatives;
```

Comment:
`derivatives` will be (n,1) vector, containing only 1s. Why do we need to multiply by a vector only containing 1s?

Reply:
The multiplication is for summing up the derivatives in each column of the matrix, where each column corresponds to a piece of testing data. I thought maybe this is faster than doing it with a for loop.